lib/Dialect/TritonGPU/IR/LinearLayoutConversions.cpp

```diff
@@ -1221,9 +1221,19 @@ LinearLayout toLinearLayout(MemDescType type) {
   // Pass in the allocation shape. Then when using invertAndCompose it will
   // trim the allocationShape to the shape if they are different.
   // We also remove the first dimension of the allocationShape if there was a
-  // call to memdesc_index
-  auto shape = type.getAllocShape().take_back(type.getRank());
-  return toLinearLayout(shape, type.getEncoding());
+  // call to memdesc_index.
+  auto shape = type.getAllocShape();
+  Attribute layout = type.getEncoding();
+  if (auto enc = dyn_cast<LayoutEncodingTrait>(layout)) {
+    unsigned rank = enc.getRank();
+    if (shape.size() > rank)
+      shape = shape.take_back(rank);
+  } else if (isa<TensorMemoryEncodingAttr, TensorMemoryScalesEncodingAttr>(
+                 layout)) {
+    if (shape.size() > 2)
+      shape = shape.take_back(2);
```

Comment:
When does this get exercised? Is there somewhere we call `toLinearLayout` on a 3d tmem allocation?

Reply:
Not in this PR. I added it as I was taking a different route in this PR, but then I dropped that route. I left it as it's more thorough than the previous approach we had for this.

Reply:
actually the previous  code was a noop ATM so let me revert this and kill the previous code, as putting this logic here is a bit too magical